ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryPartition.java

```diff
@@ -58,6 +60,21 @@
 
   @Override
   public PartitionState getPartitionState() {
+    // If there was a change to the sealed state of replicas in the cluster manager since the last check, refresh the
+    // state of this partition. We do this to avoid querying every time this method is called, considering how
+    // update to sealed states of replicas are relatively rare.
+    long currentSealedStateChangeCounter = clusterManagerCallback.getSealedStateChangeCounter();
+    if (currentSealedStateChangeCounter > lastUpdatedSealedStateChangeCounter.get()) {
+      lastUpdatedSealedStateChangeCounter.set(currentSealedStateChangeCounter);
+      boolean isSealed = false;
+      for (AmbryReplica replica : clusterManagerCallback.getReplicaIdsForPartition(this)) {
+        if (replica.isSealed()) {
+          isSealed = true;
+          break;
+        }
+      }
+      state = isSealed ? PartitionState.READ_ONLY : PartitionState.READ_WRITE;
```

Comment:
this might cause concurrency problems and hold stale state for a long time - need some concurrency control here